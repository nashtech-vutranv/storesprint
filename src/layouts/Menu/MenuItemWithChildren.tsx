import React, {useEffect, useRef} from 'react'
import * as ReactDom from 'react-dom'
import {Link} from 'react-router-dom'
import {Collapse} from 'react-bootstrap'
import classNames from 'classnames'
import {SubMenus} from './types'
import MenuItem from './MenuItem'

const MenuItemWithChildren = ({
  item,
  linkClassName,
  subMenuClassNames,
  activeMenuItems,
  toggleMenu,
  onExpandStatus,
  open,
  idx
}: SubMenus) => {
  const ref = useRef<any>()

  const toggleMenuItem = (
    e: React.MouseEvent<HTMLAnchorElement, MouseEvent>
  ) => {
    e.preventDefault()
    if (idx || idx === 0) {
      onExpandStatus && onExpandStatus(idx)
    }

    if (toggleMenu) {
      toggleMenu(item, !open)
    }

    return false
  }

  useEffect(() => {
    const menuAppItem = ReactDom.findDOMNode(ref.current) as any
    if (menuAppItem) {
      menuAppItem.setAttribute(
        'aria-label',
        menuAppItem.getAttribute('data-menu-key')
      )
    }
  }, [ref])

  return (
    <li className={classNames('side-nav-item', {'menuitem-active': open})}>
      <Link
        ref={ref}
        to='#'
        onClick={toggleMenuItem}
        data-menu-key={item.key}
        aria-expanded={open}
        className={classNames('has-arrow', 'side-sub-nav-link', linkClassName, {
          'menuitem-active': activeMenuItems!.includes(item.key)
            ? 'active'
            : '',
        })}
      >
        {item.icon && <i className={item.icon}></i>}
        {!item.badge ? (
          <span className='menu-arrow'></span>
        ) : (
          <span
            className={classNames(
              'badge',
              'bg-' + item.badge.variant,
              'float-end',
              {
                'text-dark': item.badge.variant === 'light',
              }
            )}
          >
            {item.badge.text}
          </span>
        )}
        <span> {item.label} </span>
      </Link>
      <Collapse in={open}>
        <ul className={classNames(subMenuClassNames)}>
          {(item.children || []).map((child, index) => {
            return (
              <React.Fragment key={index.toString()}>
                {child.children && (
                  <>
                    {/* parent */}
                    <MenuItemWithChildren
                      item={child}
                      linkClassName={
                        activeMenuItems!.includes(child.key) ? 'active' : ''
                      }
                      activeMenuItems={activeMenuItems}
                      subMenuClassNames='side-nav-third-level'
                      toggleMenu={toggleMenu}
                    />
                  </>
                )}
                {!child.children && (
                  <>
                    {/* child */}
                    <MenuItem
                      item={child}
                      className={
                        activeMenuItems!.includes(child.key)
                          ? 'menuitem-active'
                          : ''
                      }
                    />
                  </>
                )}
              </React.Fragment>
            )
          })}
        </ul>
      </Collapse>
    </li>
  )
}

export default MenuItemWithChildren
