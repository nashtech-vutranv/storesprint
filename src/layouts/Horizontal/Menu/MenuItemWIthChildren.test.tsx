import {render} from '@testing-library/react'
import '@testing-library/jest-dom'
import {MemoryRouter} from 'react-router-dom'
import MenuItemWithChildren from './MenuItemWithChildren'

test('renders MenuItemWithChildren', async () => {
  const tree = render(
    <MemoryRouter>
      <MenuItemWithChildren
        item={{key: 'more', label: 'test-label'}}
        className='test-classname'
        linkClassName='test-linkClassName'
        activeMenuItems={['menuItem1, menuItem2']}
      />
    </MemoryRouter>
  )
  expect(tree).toMatchSnapshot()
})
