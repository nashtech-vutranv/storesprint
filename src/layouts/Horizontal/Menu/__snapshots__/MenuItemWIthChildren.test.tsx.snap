// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders MenuItemWithChildren 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="dropdown test-classname"
      >
        <a
          aria-expanded="false"
          aria-haspopup="true"
          class="dropdown-toggle arrow-none test-linkClassName"
          data-bs-toggle="dropdown"
          data-menu-key="more"
          href="/"
          id="more"
          role="button"
        >
          <span>
             
            test-label
             
          </span>
          <div
            class="arrow-down"
          />
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="dropdown test-classname"
    >
      <a
        aria-expanded="false"
        aria-haspopup="true"
        class="dropdown-toggle arrow-none test-linkClassName"
        data-bs-toggle="dropdown"
        data-menu-key="more"
        href="/"
        id="more"
        role="button"
      >
        <span>
           
          test-label
           
        </span>
        <div
          class="arrow-down"
        />
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
